src/mlpack/tests/main_tests/hmm_training_tests.cpp

```diff
@@ -0,0 +1,209 @@
+}
+
+// Make sure that tolerance is non negative
+BOOST_AUTO_TEST_CASE(HMMTrainToleranceNonNegative)
+{
+  std::string inputFileName = "hmm_train_obs.csv";
+  int states = 3;
+  std::string hmmType = "gaussian";
+  double tol = - 100;  // Invalid
+
+  fileExists(inputFileName);
+  SetInputParam("input_file", std::move(inputFileName));
+  SetInputParam("states", states);
+  SetInputParam("type", std::move(hmmType));
+  SetInputParam("tolerance", tol);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Make sure an error is thrown if type is something other than
+// "discrete", "gaussian" or "gmm"
+BOOST_AUTO_TEST_CASE(HMMTrainTypeTest)
+{
+  std::string inputFileName = "hmm_train_obs.csv";
+  int states = 3;
+  std::string hmmType = "some-not-supported-possibly-non-type";
+
+  fileExists(inputFileName);
+  SetInputParam("input_file", std::move(inputFileName));
+  SetInputParam("states", states);
+  SetInputParam("type", std::move(hmmType));
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Make sure that the number of gaussians cannot be less than 0
+BOOST_AUTO_TEST_CASE(HMMTrainGaussianTest)
+{
+  std::string inputFileName = "hmm_train_obs.csv";
+  int states = 3;
+  std::string hmmType = "gmm";
+  int gaussians = -2;
+
+  fileExists(inputFileName);
+  SetInputParam("input_file", std::move(inputFileName));
+  SetInputParam("states", states);
+  SetInputParam("type", std::move(hmmType));
+  SetInputParam("gaussians", gaussians);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Make sure that model reuse is possible and work properly
+BOOST_AUTO_TEST_CASE(HMMTrainReuseModelTest)
+{
+  std::string inputObsFileName = "hmm_train_obs.csv";
+  std::string inputLabFileName = "hmm_train_lab.csv";
+  std::string hmmType = "discrete";
+  int states = 3;
+
+  fileExists(inputObsFileName);
+  fileExists(inputLabFileName);
+  // Make sure that the size of the 
+  // training seq, and training labels is same
+  arma::mat trainObs, trainLab;
+  data::Load(inputObsFileName, trainObs);
+  data::Load(inputLabFileName, trainLab);
+  BOOST_REQUIRE_EQUAL(trainObs.n_rows, trainLab.n_rows);
+  
+  SetInputParam("input_file", std::move(inputObsFileName));
+  SetInputParam("labels_file", std::move(inputLabFileName));
+  SetInputParam("type", std::move(hmmType));
+  SetInputParam("states", states);
+
+  mlpackMain();
+
+  HMMModel * ph1 = CLI::GetParam<HMMModel*>("output_model");
+  HMMModel h1 = *(CLI::GetParam<HMMModel*>("output_model"));
+
+  SetInputParam("input_model", std::move(ph1));
+
+  CLI::GetSingleton().Parameters()["type"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["states"].wasPassed = false;
+
+  mlpackMain();
+
+  HMMModel h2 = *(CLI::GetParam<HMMModel*>("output_model"));
```

Comment:
A minor note, but you could avoid the copy here by just working with the pointer directly.  Since this is a test on a small dataset it doesn't really matter, but something to keep in mind in any case.

Reply:
I think that atleast one copy is necessary. Consider the following
```c++
// Set some params
SetInputParam(...)

mlpackMain()

// Get the pointer to output model
HMMModel * ph1 = CLI::GetParam<HMMModel*>("output_model");

// Change some params
CLI::GetSingleton().Parameters()["..."].wasPassed = false;
SetInputParam(...)

mlpackMain();

// Again get the pointer to output model. But now the old output model is overwritten!
// Hence ph1 and ph2 both point to the same (latest) output model
HMMModel * ph2 = CLI::GetParam<HMMModel*>("output_model");
// Since we've lost out old output model, we cannot compare
```
Sorry if I'm missing something obvious

Reply:
Yep, you're right, but what I was focusing on was the copy because `h2` is not a pointer:

```
HMMModel h2 = *(CLI::GetParam<HMMModel*>("output_model"));
```

Since we are not running `mlpackMain()` again, we could just use a pointer to the second model to avoid copying the `HMMModel`:

```
HMMModel* h2 = CLI::GetParam<HMMModel*>("output_model");
```

or even use std::move():

```
HMMModel h2 = std::move(*(CLI::GetParam<HMMModel*>("output_model")));
```

This is a pretty minor comment in any case, especially since `HMMModel` only holds a few pointers so the cost of a copy is very small.  But for implementing other tests or other mlpack code, it might be something worthwhile to keep in mind. :)